Add support for Docling-based scraper

movinam opened the pull request:
### Summary :
This PR adds supports for DoclingPDFParser and Paper's base model

### Details
1. The parser can be initialised with mode as "accurate" or "fast" which enables either accurate or fast table extraction.
2. The export type can be set to either markdown or html content to parse either.
3. Add paper, section and subsections to structures.py.



# Usage


```python
from akd.tools.scrapers.web_scrapers import (
    DoclingScraper,
    DoclingScraperConfig
    WebpageScraperToolInputSchema
    SimpleWebScraper,
    Crawl4AIWebScraper
)



from akd.tools.scrapers.compoiste import CompositeScraper


_cfg_docling = DoclingScraperConfig(
    pdf_mode="fast",
    analyze_image=False,
    do_table_structure=False,
    use_ocr=False,
    debug=False,
)

_scrapers = [
    DoclingScraper(_cfg_docling),
    Crawl4AIWebScraper(),
    SimpleWebScraper(),
]

_scraper = CompositeScraper(*_scrapers)

_url = "https://arxiv.org/pdf/2402.01822v1"
# _url = "https://nishparadox.com/writing/vipassana-retreat.html"

output = await _scraper.arun(WebpageScraperToolInputSchema(url=_url)))

```

movinam committed:
Add support for DoclingPDFParser

movinam committed:
Add paper and section structures

NISH1001 commented on akd/tools/scrapers/pdf_scrapers.py:
```diff
@@ -242,3 +248,140 @@ async def arun(
+        return markdown_content, metadata
+    
+
+    async def arun(
```
With the new abstraction `_arun` is the abstract method now.

jbrry commented on akd/tools/scrapers/pdf_scrapers.py:
```diff
@@ -242,3 +248,140 @@ async def arun(
+        """
+        path = str(params.url)
+        pdf_path = str(params.url)
+        if not self._is_pdf(path):
```
What happens if this is html? `_process_pdf` permits both html and pdf

NISH1001 replied:
I think this only handles pdf right now. NEed to improve this PR

movinam replied:
_process_pdf supports exporting the pdf to HTML / markdown to further extract the content. This option is available as the extracted content varies based on the export_type. The default export type is markdown.

 It currently does not support using Docling to extract HTML content from webpages since this acts as a PDFParser.

muthukumaranR commented on akd/tools/scrapers/pdf_scrapers.py:
```diff
@@ -242,3 +248,140 @@ async def arun(
+            allowed_formats=[
+                    InputFormat.PDF,
+                    InputFormat.IMAGE,
+                ],
```
Since Docling can handle both HTML and PDF documents. Should we have a generic scrapers.py and add `DoclingParser` there?

NISH1001 replied:
Ah yes. We need to revisit this pr with new abstraction of base existing in other components.

NISH1001 replied:
cc: @iamsims check this out

NISH1001 reviewed (commented):
Need to improve this PR to adjust for new abstraction we have at `develop` branch

movinam committed:
Merge branch 'develop' of https://github.com/NASA-IMPACT/accelerated-discovery into feature/docling-scraper

movinam committed:
Replaced arun with _arun

movinam committed:
Removed duplicate code

NISH1001 commented on akd/tools/scrapers/pdf_scrapers.py:
```diff
@@ -245,3 +251,133 @@ async def _arun(
+        if not self._is_pdf(path):
+            raise RuntimeError(f"{path} is not a valid pdf.")
+        try:
+            markdown_content, metadata = await self._process_pdf(path)
```
We need to process not just pdf, HTML as well. So need to modify this if it matters.

iamsims committed:
Fix for multiple formats compatibility and code compatibility

- Add support for formats other than just pdf
- Based on WebPageScraperTool
- Make arun workable as entrypoint: threw errors in previous versions
- Configureable image based extraction and table structure detection for pdf

iamsims committed:
Remove the unnecessary exception handling

NISH1001 commented on akd/tools/scrapers/web_scrapers.py:
```diff
@@ -460,3 +464,131 @@ async def _arun(
+        export_type: str = "markdown",
+        analyze_image: bool = False,
+        do_table_structure: bool = True,
+    ):
```
This should be based off config. not parameter.

```python
from akd.tools._base import BaseToolConfig

class DoclingScraperConfig(BaseToolConfig):
...
...
```

NISH1001 commented on akd/tools/scrapers/web_scrapers.py:
```diff
@@ -460,3 +464,131 @@ async def _arun(
+class DoclingScraper(WebScraperToolBase):
+    """
+    Scrapes PDF, HTML, docs (and other supported formats) into Markdown using Docling.
+    """
```
Wehere are the IO schema? especially docling should have new inherited config, hence `config_schema = DoclingScraperConfig`

NISH1001 commented on akd/tools/scrapers/web_scrapers.py:
```diff
@@ -460,3 +464,131 @@ async def _arun(
+            InputFormat.XLSX,
+            InputFormat.ASCIIDOC,
+            InputFormat.MD,
+        ]
```
Why is this not part of config? This should be part of config and default to these formats.

NISH1001 commented on akd/tools/scrapers/web_scrapers.py:
```diff
@@ -460,3 +464,131 @@ async def _arun(
+        return markdown, metadata
+
+    async def _arun(
+        self, params: WebpageScraperToolInputSchema
```
`**kwargs` is missing from the parameter

NISH1001 committed:
Refactor docling scraper to use parameters from config

- Use `DoclingScraperConfig`

NISH1001 committed:
Move allowed formats for doclingscraper to config

- Also add image option based on `analyze_image` value

NISH1001 committed:
[Chore] Add kwargs arugment to _arun in doclingscraper

- Added it to conform to linting for base

NISH1001 committed:
Merge branch 'develop' into feature/docling-scraper

NISH1001 committed:
Add ocr option in doclingscraper config

Default to false with `do_ocr` paramter

NISH1001 committed:
Merge branch 'feature/docling-scraper' of github.com:NASA-IMPACT/accelerated-discovery into feature/docling-scraper

NISH1001 committed:
Remove uwnanted imports from pdf parser module

Also add debug log for docling

NISH1001 committed:
Add htmlformat option for docling scraper

NISH1001 commented on akd/tools/scrapers/web_scrapers.py:
```diff
@@ -460,3 +469,167 @@ async def _arun(
+        metadata = WebpageMetadata(
+            url=path,
+            query=path,
+            title="Untitled",
```
THis nees to get resolved before we merge. There has to be some simple heuristics to determine this.

NISH1001 committed:
Extract title from docling document

- Use `doc.texts` contents
- If not found, ust `doc.name` directly
- Else return untitled

NISH1001 committed:
Improve title extraction from docling document

- Use SectionHeaderItem
- Use TitleItem

NISH1001 committed:
Improve title extraction for docling scraper

- Use `scrapers.utils._DoclingTitleExtractor` for conveience

NISH1001 committed:
Rename docling title extractor to metadata extractor

NISH1001 committed:
Remove unwanted code block from docling scraper

NISH1001 committed:
Merge branch 'develop' of github.com:NASA-IMPACT/accelerated-discovery into feature/docling-scraper
